File: ExeProcessorPlus.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.util.Vector;

enum ExeProcessorType {
	PRINT, EXPORT, OTHERS
}

public class ExeProcessorPlus {
	private String exePath;
	private Process process = null;
	private Vector<String> argvs;

	public String getExePath() {
		return exePath;
	}

	public void setExePath(String exePath) {
		this.exePath = exePath;
	}

	public Vector<String> getCommandStr() {
		return argvs;
	}

	public void setCommandStr(Vector<String> commandStr) {
		this.argvs = commandStr;
	}

	private void dealPipeData(String data, ExeProcessorType type, BufferedWriter writer) {
		try {
			switch (type) {
			case PRINT:
				System.out.println(data);
				break;
			case EXPORT:
				writer.write(data);
				writer.newLine();
				break;
			case OTHERS:
				break;
			}
		} catch (Exception e) {
			// TODO: handle exception
			e.printStackTrace();
		}
	}

	private void internalExecute(ExeProcessorType type,String exportPath) {
		try {
			// 内部使用对象，来完成可能的对外txt文件输出
			File dstFile = null;
			FileWriter dstFileWriter = null;
			BufferedWriter dstbWriter = null;

			if (type == ExeProcessorType.EXPORT) {
				dstFile = new File(exportPath);
				dstFile.createNewFile();
				dstFileWriter = new FileWriter(dstFile);
				dstbWriter = new BufferedWriter(dstFileWriter);
			}

			// 从管道中读出输出信息
			if (dstbWriter != null) {
				dstbWriter.write("输出信息");
				dstbWriter.newLine();
				dstbWriter.newLine();
			} else {
				System.out.println("输出信息");
			}
			BufferedReader stdOut = new BufferedReader(new InputStreamReader(process.getInputStream()));

			for (String line; null != (line = stdOut.readLine());)
				dealPipeData(line, type, dstbWriter);

			// 从管道中读错误信息
			if (dstbWriter != null) {
				dstbWriter.write("错误信息");
				dstbWriter.newLine();
				dstbWriter.newLine();
			} else {
				System.out.println("错误信息");
			}

			BufferedReader stdError = new BufferedReader(new InputStreamReader(process.getErrorStream()));

			for (String line; null != (line = stdError.readLine());)
				dealPipeData(line, type, dstbWriter);

			if (dstbWriter != null) {
				dstbWriter.close();
			}

			if (dstFileWriter != null) {
				dstFileWriter.close();
			}

			if (stdOut != null) {
				stdOut.close();
			}

			if (stdError != null) {
				stdError.close();
			}

		} catch (Exception e) {
			// TODO: handle exception
			e.printStackTrace();
		}
	}

	public boolean executeRun(String path, Vector<String> argvs, ExeProcessorType type, String exportPath,
			boolean isParallel) {
		try {
			if (path.contains(" ")) {
				path = "\"" + path + "\"";
			}

			this.setExePath(path);
			this.setCommandStr(argvs);
			String cmdStr = exePath;
			for (int i = 0; i < argvs.size(); i++) {
				cmdStr = cmdStr + " " + argvs.get(i);
			}

			System.out.println(cmdStr);

			this.process = Runtime.getRuntime().exec(cmdStr);

			if (!isParallel) {
				this.process.waitFor();
			}

			// 向管道中输入数据
			/*
			 * new Thread() { public void run() { OutputStream stdin =
			 * process.getOutputStream(); for (int i = 0; stdin!=null ; i++) { try {
			 * Thread.sleep(1); // 要休息片刻才看得到 I/O 的缓存效果。 stdin.write((i + " " + i +
			 * "\n").getBytes()); } catch (Exception ex) { ex.printStackTrace(); } } }
			 * }.start();
			 */
			
			if (isParallel) {
				new Thread() {
					public void run() {
						internalExecute(type, exportPath);
					};
				}.start();
			}
			else {
				internalExecute(type, exportPath);
			}
			

		} catch (Exception e) {
			// TODO: handle exception
			return false;
		}

		return true;
	}

	public static void main(String[] args) {
		// TODO 自动生成的方法存根
		ExeProcessorPlus exeProcessorPlus = new ExeProcessorPlus();
		Vector<String> argvs = new Vector<String>();
		argvs.add("123");
		argvs.add("456");

		exeProcessorPlus.executeRun("exeDemo/test.exe", argvs, ExeProcessorType.PRINT, "", true);

		while (true) {
			System.out.println("这是主线程");
		}
	}

}

File: ExeProcessorPlus.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Diagnostics;
using System.IO;

namespace ExeProcessor
{
    public enum ExeProcessorType
    {
        PRINT, EXPORT, OTHERS
    }


    class ExeProcessorPlus
    {

        //Windows进入cmd命令专用
        public static String CMD_INTRICATOR = "cmd.exe ";
        //Windows系统单次多条命令顺序执行连接符
        public static String EXE_INTRICATOR_SEPERATE = " & ";


        //Exe的路径设置
        private String exePath="";

        public String ExePath
        {
            get { return exePath; }
            set { exePath = value; }
        }

        private Process process = null;

        //传入参数
        private List<String> argvs=null;
        public List<String> Argvs
        {
            get { return argvs; }
            set { argvs = value; }
        }

        ExeProcessorPlus()
        {
            ExePath = "";
            process = null;
            if (Argvs!=null)
            {
                Argvs.Clear();
            }
        }

        //对数据的方法判定
        private void dealPipeData(String data, ExeProcessorType type, StreamWriter dstWriter)
        {
            try
            {
                switch (type)
                {
                    case ExeProcessorType.PRINT:
                        Console.WriteLine(data);
                        break;
                    case ExeProcessorType.EXPORT:
                        dstWriter.WriteLine(data);
                        break;
                    case ExeProcessorType.OTHERS:
                        //自定义处理方法


                        break;
                }
            }
            catch (System.Exception ex)
            {
                Console.WriteLine(ex.Data);
            }
        }

        //对输出流和错误流是否与主线程并行的处理方法
        private void internalExecute(ExeProcessorType type, String exportPath, Process exeProcessor)
        {
            try
            {
                StreamWriter dstWriter = null;

                //从管道中输出信息
                if (type == ExeProcessorType.EXPORT)
                {
                    dstWriter = new StreamWriter(exportPath);
                }

                if (dstWriter != null)
                {
                    dstWriter.WriteLine("输出信息");
                    dstWriter.WriteLine();
                }
                else
                {
                    Console.WriteLine("输出信息");
                }

                while (!exeProcessor.StandardOutput.EndOfStream)
                {
                    String line = exeProcessor.StandardOutput.ReadLine();
                    dealPipeData(line, type, dstWriter);
                }

                //从管道中读错误信息
                if (dstWriter != null)
                {
                    dstWriter.WriteLine("错误信息");
                    dstWriter.WriteLine();
                }
                else
                {
                    Console.WriteLine("错误信息");
                }

                while (!exeProcessor.StandardError.EndOfStream)
                {
                    String line = exeProcessor.StandardError.ReadLine();
                    dealPipeData(line, type, dstWriter);
                }

                if (dstWriter != null)
                {
                    dstWriter.Close();
                }
            }
            catch (System.Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        private String getArgvStr(List<String> argvs)
        {
            Argvs = argvs;
            String argvStr = "";
            for (int i = 0; i < Argvs.Count; i++)
            {
                argvStr =argvStr+ Argvs[i];
                if (i != Argvs.Count - 1)
                {
                    argvStr = argvStr + " ";
                }
            }
            return argvStr;
        }

        //对外封装的执行exe方法
        public bool executeRun(String path, List<String> argvs, ExeProcessorType type,
            String exportPath,bool isParallel)
        {
            try
            {
                ExePath = path;
                String argvStr = getArgvStr(argvs);

                //输出最终的指令
                Console.WriteLine(ExePath + " " + argvStr);

                //声明Exe调用对象
                Process exeProcessor = new Process();
                //设置Exe路径和输入参数
                exeProcessor.StartInfo.FileName = ExePath;
                exeProcessor.StartInfo.Arguments = argvStr;
                exeProcessor.StartInfo.CreateNoWindow = true;
                exeProcessor.StartInfo.UseShellExecute = false;


                //输出流和错误流设置
                exeProcessor.StartInfo.RedirectStandardOutput = true;
                exeProcessor.StartInfo.RedirectStandardError = true;
                //输入流设置
                exeProcessor.StartInfo.RedirectStandardInput = true;
                //p.StandardInput.AutoFlush = true;

                exeProcessor.Start();

                if (!isParallel)
                {
                    internalExecute(type, exportPath, exeProcessor);
                }
                else
                {
                    Thread parallelThread = new Thread(delegate()
                    {
                        internalExecute(type, exportPath, exeProcessor);
                    });
                    parallelThread.Start();
                }
            }
            catch (System.Exception ex)
            {
                Console.WriteLine(ex.Message);
            }

            return true;
        }

        //对外封装的执行exe方法
        public bool executeRunForWindows(List<String> prefixGroup,String path, List<String> argvs, ExeProcessorType type,
            String exportPath, bool isParallel)
        {
            try
            {
                ExePath = path;
                String cmdStr = "";
                foreach (var item in prefixGroup)
                {
                    cmdStr = cmdStr + item + EXE_INTRICATOR_SEPERATE;
                }

                cmdStr=cmdStr+ExePath+" "+getArgvStr(argvs);

                //输出最终的指令
                Console.WriteLine(cmdStr);

                //声明Exe调用对象
                Process exeProcessor = new Process();
                //设置Exe路径和输入参数，这里是要先调用cmd.exe再去把其他的当作传参的
                exeProcessor.StartInfo.FileName = CMD_INTRICATOR;
                exeProcessor.StartInfo.CreateNoWindow = true;
                exeProcessor.StartInfo.UseShellExecute = false;


                //输出流和错误流设置
                exeProcessor.StartInfo.RedirectStandardOutput = true;
                exeProcessor.StartInfo.RedirectStandardError = true;
                //输入流设置
                exeProcessor.StartInfo.RedirectStandardInput = true;
                //p.StandardInput.AutoFlush = true;

                exeProcessor.Start();

                exeProcessor.StandardInput.WriteLine(cmdStr);
                exeProcessor.StandardInput.AutoFlush = true;

                if (!isParallel)
                {
                    internalExecute(type, exportPath, exeProcessor);
                }
                else
                {
                    Thread parallelThread = new Thread(delegate()
                    {
                        internalExecute(type, exportPath, exeProcessor);
                    });
                    parallelThread.Start();
                }
            }
            catch (System.Exception ex)
            {
                Console.WriteLine(ex.Message);
            }

            return true;
        }

        static void Main(string[] args)
        {
            ExeProcessorPlus exeRuntime = new ExeProcessorPlus();
            
            //方式一:直接执行对应的Exe
            List<String> argvs = new List<string>();
            argvs.Add("123");
            argvs.Add("456");
            String exePath="./exeDemo/test.exe";
            exeRuntime.executeRun(exePath, argvs, ExeProcessorType.PRINT, "", true);

            //方式二：切换到对应的环境下再执行对应的EXE(针对Windows系统),为了保证有些局部变量能被正常的使用
            //List<String> prefixGroups = new List<String>();
            //prefixGroups.Add("f:");
            //prefixGroups.Add("cd f:");
            //String exePath = "test.exe";

            //List<String> argvs = new List<String>();
            //argvs.Add("1545446");
            //argvs.Add("5959595");
            //exeRuntime.executeRunForWindows(prefixGroups, exePath, argvs, ExeProcessorType.PRINT, "",true);

            while (true)
            {
                Console.WriteLine("这是主线程");
            }

            Console.ReadLine();
        }
    }
}
